File: calculadora_programador.py
#!/usr/bin/env python3
"""
Calculadora do Programador (Bases & Bitwise)
Conversão simultânea entre Decimal, Hexadecimal, Binário e Octal,
mais operações lógicas bit a bit (AND, OR, XOR, NOT, Shift).
Execução: python calculadora_programador.py
"""

import sys


class CalculadoraProgramador:
    def __init__(self, bits: int = 32):
        self.bits = bits
        self.mascara = (1 << bits) - 1

    def converter(self, valor: int) -> dict:
        v_trunc = valor & self.mascara
        return {
            "dec": str(valor),
            "hex": hex(v_trunc).upper().replace("X", "x"),
            "bin": bin(v_trunc)[2:].zfill(self.bits),
            "oct": oct(v_trunc),
        }

    def operacao_bitwise(self, op: str, a: int, b: int = 0) -> int:
        if op == "AND":
            return (a & b) & self.mascara
        elif op == "OR":
            return (a | b) & self.mascara
        elif op == "XOR":
            return (a ^ b) & self.mascara
        elif op == "NOT":
            return (~a) & self.mascara
        elif op == "SHL":
            return (a << b) & self.mascara
        elif op == "SHR":
            return (a >> b) & self.mascara
        else:
            raise ValueError(f"Operador bitwise '{op}' desconhecido")


def main():
    calc = CalculadoraProgramador(bits=32)
    print("=" * 50)
    print("   CALCULADORA DO PROGRAMADOR (BASES & BITS)   ")
    print("=" * 50)
    print("Exemplos de entrada de valor:")
    print("  Decimal: 255  |  Hex: 0xFF  |  Bin: 0b11111111\n")

    while True:
        try:
            entrada = input("Digite um número ou expressão (ou 'sair'): ").strip()
        except (KeyboardInterrupt, EOFError):
            break

        if not entrada or entrada.lower() in ["sair", "exit"]:
            break

        try:
            # Reconhece prefixos 0x, 0b, 0o ou int comum
            num = int(entrada, 0)
            res = calc.converter(num)
            print(f"\n--- Representação em Múltiplas Bases ({calc.bits} bits) ---")
            print(f"  DEC: {res['dec']}")
            print(f"  HEX: {res['hex']}")
            print(f"  OCT: {res['oct']}")
            # Agrupa binário em nibbles de 4 bits para legibilidade
            bin_fmt = " ".join(res["bin"][i:i+4] for i in range(0, len(res["bin"]), 4))
            print(f"  BIN: {bin_fmt}\n")
        except Exception as err:
            print(f"Erro: {err}\n")


if __name__ == "__main__":
    main()

File: CalculadoraProgramador.java
import java.util.Scanner;

/**
 * CalculadoraProgramador.java
 * Calculadora do Programador em Java: Bases (HEX, DEC, OCT, BIN) e Operadores Bitwise.
 * Execução: java CalculadoraProgramador.java
 */
public class CalculadoraProgramador {

    private int bits = 32;

    public void exibirRelatorio(long valor) {
        long mascara = bits == 64 ? -1L : (1L << bits) - 1;
        long valorMascarado = valor & mascara;

        System.out.println("=================================================");
        System.out.printf("   CONVERSÃO DE BASES PARA: %d (%d bits)%n", valor, bits);
        System.out.println("=================================================");
        System.out.printf("   DEC: %d%n", valorMascarado);
        System.out.printf("   HEX: 0x%s%n", Long.toHexString(valorMascarado).toUpperCase());
        System.out.printf("   OCT: 0%s%n", Long.toOctalString(valorMascarado));

        String binStr = Long.toBinaryString(valorMascarado);
        while (binStr.length() < bits) binStr = "0" + binStr;
        if (binStr.length() > bits) binStr = binStr.substring(binStr.length() - bits);

        // Agrupando em nibbles (4 bits)
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < binStr.length(); i++) {
            if (i > 0 && i % 4 == 0) sb.append(" ");
            sb.append(binStr.charAt(i));
        }
        System.out.printf("   BIN: %s%n", sb.toString());
        System.out.printf("   BITS ATIVOS (popcount): %d%n", Long.bitCount(valorMascarado));
        System.out.println("-------------------------------------------------");
        System.out.println(" OPERAÇÕES BITWISE COM MÁSCARA 0x0F (15):");
        System.out.printf("   AND (& 15): %d (0x%X)%n", (valorMascarado & 15), (valorMascarado & 15));
        System.out.printf("   OR  (| 15): %d (0x%X)%n", (valorMascarado | 15), (valorMascarado | 15));
        System.out.printf("   XOR (^ 15): %d (0x%X)%n", (valorMascarado ^ 15), (valorMascarado ^ 15));
        System.out.printf("   NOT (~):    %d (0x%X)%n", (~valorMascarado & mascara), (~valorMascarado & mascara));
        System.out.printf("   SHL (<< 2): %d%n", ((valorMascarado << 2) & mascara));
        System.out.printf("   SHR (>> 2): %d%n", (valorMascarado >> 2));
        System.out.println("=================================================");
    }

    public static void main(String[] args) {
        CalculadoraProgramador prog = new CalculadoraProgramador();

        if (args.length > 0) {
            for (String arg : args) {
                try {
                    long val = arg.startsWith("0x") || arg.startsWith("0X")
                        ? Long.parseLong(arg.substring(2), 16)
                        : Long.parseLong(arg);
                    prog.exibirRelatorio(val);
                } catch (NumberFormatException e) {
                    System.err.println("Número inválido: " + arg);
                }
            }
            return;
        }

        System.out.println("Demonstração padrão da Calculadora do Programador:");
        long[] amostras = { 42, 255, 1024, 65535 };
        for (long a : amostras) {
            prog.exibirRelatorio(a);
            System.out.println();
        }
    }
}
